Migrate emoncms to config flow

alexandrecuer opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change

Migrate emoncms to config flow

**A PR to modify the documentation is to follow**


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [x] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request:https://github.com/home-assistant/home-assistant.io/pull/34554

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] I have followed the [perfect PR recommendations][perfect-pr]
- [X] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [X] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [X] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

alexandrecuer committed:
Migrate emoncms to config flow

mib1185 commented on homeassistant/components/emoncms/manifest.json:
```diff
@@ -2,7 +2,8 @@
   "documentation": "https://www.home-assistant.io/integrations/emoncms",
   "iot_class": "local_polling",
-  "requirements": ["pyemoncms==0.0.7"]
+  "requirements": ["pyemoncms==0.1.1"]
```
Is the bump of this lib necessary for the config flow implementation?

joostlek commented on homeassistant/components/emoncms/__init__.py:
```diff
@@ -1 +1,32 @@
+
+from .const import DOMAIN, LOGGER
+
+PLATFORMS: list[str] = ["sensor"]
```
Please use `Platform.SENSOR`

joostlek commented on homeassistant/components/emoncms/config_flow.py:
```diff
@@ -0,0 +1,209 @@
+        errors: dict[str, str] = {}
+
+        if user_input is not None:
+            self._async_abort_entries_match(
+                {
+                    CONF_API_KEY: user_input[CONF_API_KEY],
+                    CONF_ID: user_input[CONF_ID],
+                    CONF_URL: user_input[CONF_URL],
+                }
+            )
```
Hmm, That UUID you added recently would work perfect here

joostlek commented on homeassistant/components/emoncms/config_flow.py:
```diff
@@ -0,0 +1,209 @@
+                        vol.Required(CONF_ID): str,
+                        vol.Required(CONF_URL): str,
+                        vol.Required(CONF_API_KEY): str,
+                        vol.Required(CONF_SCAN_INTERVAL, default=30): int,
```
scan interval should not be configurable. People can use the `update_entity` service to change the way they poll

joostlek commented on homeassistant/components/emoncms/__init__.py:
```diff
@@ -1 +1,32 @@
+PLATFORMS: list[str] = ["sensor"]
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
```
We should use `__init__.py` to check if we can connect to the service and if we can continue setting up the integration

alexandrecuer committed:
tests coverage 98%

alexandrecuer committed:
use runtime_data

alexandrecuer committed:
Remove pyemoncms bump.

joostlek commented on homeassistant/components/emoncms/config_flow.py:
```diff
@@ -0,0 +1,203 @@
+            data_schema=self.add_suggested_values_to_schema(
+                vol.Schema(
+                    {
+                        vol.Required(CONF_ID): str,
```
The ID is not needed to connect? It currently looks like a `name` field, but we don't allow users to set their own name unless its a helper

alexandrecuer replied:
Not the id is not needed to connect, yes it is just a helper, you can modify after with the rename facility dedicated to the config entry
![image](https://github.com/user-attachments/assets/9782cbab-fe95-456d-afa9-8adba8a44398)
You are right, it is something like a name field

joostlek replied:
Yea so let's remove this and set the config entry title to something the user can recognize

alexandrecuer replied:
done

joostlek commented on homeassistant/components/emoncms/config_flow.py:
```diff
@@ -0,0 +1,203 @@
+        value_template = None
+        exclude_feeds = None
+        include_only_feeds = None
+        if import_info.get(CONF_VALUE_TEMPLATE) is not None:
+            value_template = import_info[CONF_VALUE_TEMPLATE].template
+        if import_info.get(CONF_EXCLUDE_FEEDID) is not None:
+            exclude_feeds = list(map(str, import_info[CONF_EXCLUDE_FEEDID]))
+        if import_info.get(CONF_ONLY_INCLUDE_FEEDID) is not None:
+            include_only_feeds = list(map(str, import_info[CONF_ONLY_INCLUDE_FEEDID]))
```
So I think there are 2 things we need to figure out before we can continue:
1. The integration uses templates in YAML to pre process their data (if I understood correctly). We don't want that kind of feature when we move to the config flow. Integrations should expose contextualized "raw" data from the API or device. The user can then use templates and helpers and what not to tailer it to their needs.
2. Config flow integrations are one size fits all kind of deal. Where previously you could be picky with the YAML setup to only add entities which you wish. With config flows we want to keep it simple. The main config flow should just ask for the url and api key (like we do now) and just add everything. The user can then disable stuff as they wish. I think we should ignore the YAML config and just have 1 type of config entry (so we don't have 1 "new" config entry which adds everything and 1 "old" config entry which omits stuff).

alexandrecuer replied:
not really preprocessing, mostly to choose feeds they want to include and to exclude. When you have 100 feeds imported in one shot and you want to keep only 10, it is a lot of work for the user to remove the 90 feeds he does not want to keep. So I thought it more friendly not to import nothing at the entry creation, just to test that the connection and then when the user click on configure he chooses the feeds he wants to include....

I agree it is better to remove the `value_template`

joostlek replied:
Can you give more context about how many feeds an average instance has?

alexandrecuer replied:
60 or 100 feeds is something usual. @borpin, can you confirm or give further details?

borpin replied:
@joostlek - there is no typical use. As a minimum, on an older system 6-10 available Feeds, at extremis 1000 (or more and everything in between). I have about 300 in one instance, but a lot of these are actually coming *from* HA (so would not want them back again). There also may be historical feeds that are no longer updated, but kept.

There is no limit on the number Feeds that can be created.

One node/tag
![image](https://github.com/user-attachments/assets/f424fc4a-7f59-4273-b7ee-8fd251ef4ddc)

@alexandercuer - Can you do the Feed selection by Node, then feed?

I would agree that you only want the user to pull in the Feeds they want, allowing the user to select the feeds required (as per the current design). A 1000 disabled feeds would be a pest in HA.

1. Reduce load on emoncms instance API - only pull required feeds
2. If a feed is misconfigured, you do not want to be loading it into HA.
3. Minimise the list of sensors available to the user.

I'd also say, try and keep the sensor name compact. I'd suggest <instance>-<node>-<feed> - (I've not kept up with what you are proposing), but some integrations have impossibly long sensor names.

One other thought - the FeedID is only ever used once and cannot be changed. The Name/description in emoncms can be changed - could this simply pass through to HA?

alexandrecuer replied:
Only kept `CONF_ONLY_INCLUDE_FEEDID` from the YAML config, so this should be close to what you want

joostlek commented on homeassistant/components/emoncms/coordinator.py:
```diff
@@ -19,13 +18,12 @@ def __init__(
-            _LOGGER,
+            LOGGER,
             name="emoncms_coordinator",
             update_method=emoncms_client.async_list_feeds,
```
Can we move the updating method to `_async_update_data` and properly catch exceptions and raise `UpdateFailed` when that happens? This way we can omit the connection check in `__init__.py` because `coordinator.async_config_entry_first_refresh()` handles this then

alexandrecuer replied:
done

borpin commented:
@alexandrecuer - I saw a comment somewhere about only allowing one entry per URL. Because emoncms is User Based, there could be a need for more than one entry per URL, but for a different user I think.

A different user will get a different API key to distinguish them (again I think) - this is how emoncms.org works. I could have 2 accounts (uaser) for different installations that I want to monitor in HA. So would need 2 separate instances for the same URL.

I may have misunderstood the comment though (I can't find it again).

alexandrecuer commented:
> @alexandrecuer - I saw a comment somewhere about only allowing one entry per URL. Because emoncms is User Based, there could be a need for more than one entry per URL, but for a different user I think.
> 
> A different user will get a different API key to distinguish them (again I think) - this is how emoncms.org works. I could have 2 accounts (users) for different installations that I want to monitor in HA. So would need 2 separate instances for the same URL.
> 
> I may have misunderstood the comment though (I can't find it again).

Normally this is OK. It is one entry for a pair (url, apikey) = one entry per emoncms uuid.

alexandrecuer committed:
Remove not needed yaml parameters add async_update_data to coordinator

borpin commented:
> one entry per emoncms uuid

And this is by User? (again I've lost track a bit :( )

alexandrecuer commented:
> > one entry per emoncms uuid
> 
> And this is by User? (again I've lost track a bit :( )

Yes the uuid in emoncms is one for each user account

alexandrecuer committed:
Reduce snapshot size

alexandrecuer committed:
Merge branch 'dev' into config_flow

alexandrecuer committed:
Merge branch 'dev' into config_flow

alexandrecuer committed:
Remove CONF_UNIT_OF_MEASUREMENT

alexandrecuer committed:
Merge branch 'dev' into config_flow

alexandrecuer committed:
correct path in emoncms_client mock

alexandrecuer committed:
Remove init connexion check
as done by config_entry_first_refresh
since async_update_data catches exceptionand raise UpdateFailed

alexandrecuer committed:
Remove CONF_EXCLUDE_FEEDID from config flow

joostlek commented:
I am also wondering, if there are 1000 possible things to import. Is there a way we can create a division based on how important each thing is? So we could disable the rest that is not important but allow to add it with ease

alexandrecuer commented:
> I am also wondering, if there are 1000 possible things to import. Is there a way we can create a division based on how important each thing is? So we could disable the rest that is not important but allow to add it with ease

You mean if we identify more than 1000 feeds available for example, limit the import to only 100 or to a maximum limit we can fix in the code ? There's no way to specify the importance of certain feeds over others in emoncms. The good thing is to give the user the choice of importing what he wants at any time, which is in fact done.

alexandrecuer commented:
Other solution : import automatically  all feeds with a power or energy unit.

joostlek commented:
What other options are there? I'm consulting some others about this but I'm on the campsite so it's a little slower

borpin commented:
> Other solution : import automatically all feeds with a power or energy unit.

No, I have lots of feeds with Power and Energy I would not want imported automatically.

> The good thing is to give the user the choice of importing what he wants at any time, which is in fact done.

For this integration, this is the best option - to allow the user to select what they want to import. Few other integrations would have the potential to import such a high number of potential sensors.

This is in effect a database integration - allowing the 'pull' of certain data records within the database. (we can deal with push later 😁 )

Compare to the SQL integration that allows you to select the data to be 'pulled' in by way of queries.

joostlek commented on homeassistant/components/emoncms/__init__.py:
```diff
@@ -1 +1,38 @@
+PLATFORMS: list[str] = [Platform.SENSOR]
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
```
Let's extend the ConfigEntry type to type the runtime_data_field

alexandrecuer replied:
done, if my understanding is correct...

joostlek commented on homeassistant/components/emoncms/__init__.py:
```diff
@@ -1 +1,38 @@
+
+from .coordinator import EmoncmsCoordinator
+
+PLATFORMS: list[str] = [Platform.SENSOR]
```
```suggestion
PLATFORMS: list[Platform] = [Platform.SENSOR]
```

joostlek commented on homeassistant/components/emoncms/config_flow.py:
```diff
@@ -0,0 +1,170 @@
+        url = self._config_entry.data[CONF_URL]
+        api_key = self._config_entry.data[CONF_API_KEY]
+        include_only_feeds = self._config_entry.data.get(CONF_ONLY_INCLUDE_FEEDID, [])
+        options: Any = include_only_feeds
```
What is this statement for?

alexandrecuer replied:
to initialize the options used by the select

joostlek replied:
No I mean why do we type it as `Any`

alexandrecuer replied:
done

joostlek commented on homeassistant/components/emoncms/config_flow.py:
```diff
@@ -0,0 +1,170 @@
+            api_key = user_input[CONF_API_KEY]
+            include_only_feeds = user_input[CONF_ONLY_INCLUDE_FEEDID]
+            if self.hass.config_entries.async_update_entry(
+                self._config_entry, title=sensor_name(url), data=user_input
```
Don't update the title

joostlek commented on homeassistant/components/emoncms/config_flow.py:
```diff
@@ -0,0 +1,170 @@
+            url = user_input[CONF_URL]
+            api_key = user_input[CONF_API_KEY]
+            include_only_feeds = user_input[CONF_ONLY_INCLUDE_FEEDID]
+            if self.hass.config_entries.async_update_entry(
+                self._config_entry, title=sensor_name(url), data=user_input
+            ):
+                LOGGER.debug("entry updated")
+            return self.async_create_entry(data=user_input)
```
We should store this in the options instead, currently if you would submit the options flow it would reload the integration twice. So please only use the `self.async_create_entry`

alexandrecuer replied:
done

joostlek commented on homeassistant/components/emoncms/config_flow.py:
```diff
@@ -0,0 +1,170 @@
+                    vol.Required(CONF_API_KEY, default=api_key): str,
+                    vol.Required(
+                        CONF_ONLY_INCLUDE_FEEDID, default=include_only_feeds
+                    ): selector({"select": dropdown}),
```
What's the "select" for?

alexandrecuer replied:
the select permits the user to choose the feeds he wants to include

joostlek commented on homeassistant/components/emoncms/sensor.py:
```diff
@@ -76,22 +80,45 @@ async def async_setup_platform(
+            f"deprecated_yaml_{DOMAIN}",
+            is_fixable=False,
+            issue_domain=DOMAIN,
+            breaks_in_ha_version="2025.2.0",
```
```suggestion
            breaks_in_ha_version="2025.3.0",
```

joostlek commented on homeassistant/components/emoncms/sensor.py:
```diff
@@ -76,22 +80,45 @@ async def async_setup_platform(
-    emoncms_client = EmoncmsClient(url, apikey, session=async_get_clientsession(hass))
-    coordinator = EmoncmsCoordinator(hass, emoncms_client, scan_interval)
+    coordinator = entry.runtime_data
     await coordinator.async_refresh()
```
Not needed as the coordinator already has data

joostlek commented on homeassistant/components/emoncms/sensor.py:
```diff
@@ -146,20 +159,9 @@ def __init__(
         self._attr_native_unit_of_measurement = unit_of_measurement
-        self._sensorid = sensorid
-
+        self._attr_unique_id = f"{entry_id}-{elem[FEED_ID]}"
```
Double check: previous sensors didnt have unique ids?

joostlek replied:
Yea I can remember unique_id and emoncms PR existed, but not sure what happened with it. Thanks

joostlek commented on homeassistant/components/emoncms/strings.json:
```diff
@@ -0,0 +1,24 @@
+    "step": {
+      "user": {
+        "data": {
+          "id": "id",
```
id?

alexandrecuer replied:
removed

joostlek commented on homeassistant/components/emoncms/strings.json:
```diff
@@ -0,0 +1,24 @@
+      "user": {
+        "data": {
+          "id": "id",
+          "url": "Server Hostname",
+          "api_key": "API Key"
```
Please use references to the common strings. Use the data_description field to add more context to the fields

alexandrecuer replied:
done

joostlek commented on homeassistant/components/emoncms/strings.json:
```diff
@@ -0,0 +1,24 @@
+        "data": {
+          "url": "[%key:component::emoncms::config::step::user::data::url%]",
+          "api_key": "[%key:component::emoncms::config::step::user::data::api_key%]",
+          "include_only_feed_id": "choose feeds to include"
```
```suggestion
          "include_only_feed_id": "Choose feeds to include"
```

joostlek commented on homeassistant/components/emoncms/config_flow.py:
```diff
@@ -0,0 +1,170 @@
+            step_id="init",
+            data_schema=vol.Schema(
+                {
+                    vol.Required(CONF_URL, default=url): str,
+                    vol.Required(CONF_API_KEY, default=api_key): str,
```
Options flow should not be used for resetting the url or api key. We have a reauth or reconfigure flow for that (but don't add that in this PR, it's for a followup)

alexandrecuer replied:
url and api_key removed from options flow

joostlek reviewed (changes requested):
I also talked with the core team, with this much entities being created from the start, we've agreed on using a selector, because there is no way to limit the feeds shown in the application itself

alexandrecuer committed:
Update homeassistant/components/emoncms/__init__.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

alexandrecuer committed:
Update homeassistant/components/emoncms/sensor.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

borpin commented:
@alexandrecuer - are you still using a numeric ID for the emoncms Instance? If so, perhaps a text ID would be better, to more easily identify the related Sensors?

alexandrecuer committed:
Update homeassistant/components/emoncms/strings.json

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

joostlek commented:
What is the significance of the id anyway?

alexandrecuer commented:
> @alexandrecuer - are you still using a numeric ID for the emoncms Instance? If so, perhaps a text ID would be better, to more easily identify the related Sensors?

no the id is nor used anymore. It is still present in `strings.json` as I forgot to remove it.

alexandrecuer committed:
Use options in options flow and common strings

alexandrecuer commented on homeassistant/components/emoncms/config_flow.py:
```diff
@@ -0,0 +1,170 @@
+        include_only_feeds = None
+        if import_info.get(CONF_ONLY_INCLUDE_FEEDID) is not None:
+            include_only_feeds = list(map(str, import_info[CONF_ONLY_INCLUDE_FEEDID]))
+        if not include_only_feeds:
+            emoncms_result = await get_feed_list(self.hass, url, api_key)
+            if emoncms_result[CONF_SUCCESS]:
+                include_only_feeds = [
+                    feed[FEED_ID] for feed in emoncms_result[CONF_MESSAGE]
+                ]
```
I think I should remove these lines so that when yaml is imported, feed list is imported from the yaml only if `include_only_feed_id` is present ? These lines are used to import **all the feeds** when the user has no `include_only_feed_id` in his yaml configuration....but we dont really want this as we want the user to choose with the select so that he can import only what he really needs....

alexandrecuer replied:
done

alexandrecuer committed:
Extend the ConfigEntry type

joostlek commented on homeassistant/components/emoncms/__init__.py:
```diff
@@ -1 +1,40 @@
+
+
+async def async_setup_entry(
+    hass: HomeAssistant, entry: ConfigEntry[EmoncmsCoordinator]
```
Please define the type outside here
```py
type EmonCMSConfigEntry = ConfigEntry[EmoncmsCoordinator]
```

alexandrecuer replied:
done

joostlek commented on homeassistant/components/emoncms/config_flow.py:
```diff
@@ -0,0 +1,170 @@
+            else:
+                title = sensor_name(user_input[CONF_URL])
+                return self.async_create_entry(title=title, data=user_input)
+        return self.async_show_form(
+            step_id="user",
+            data_schema=self.add_suggested_values_to_schema(
+                vol.Schema(
+                    {
+                        vol.Required(CONF_URL): str,
+                        vol.Required(CONF_API_KEY): str,
+                    }
+                ),
+                user_input,
+            ),
```
We can now ask the user for what flows to import

alexandrecuer replied:
maybe I misunderstand something but just to be sure : you want I add something like a `async_step_choose_feeds` method in `EmoncmsConfigFlow` ? because I dont know if it is possible to use the `async_step_init` method of `EmoncmsOptionsFlow` in `EmoncmsConfigFlow`

If a new method is needed, we could keep this for a future PR don't you think ?

joostlek replied:
I am just pointing this out because it was in the original design and now I discussed this with the core team we approve this. So if you still want to add it, go ahead.

joostlek commented on homeassistant/components/emoncms/sensor.py:
```diff
@@ -146,20 +158,9 @@ def __init__(
         self._attr_native_unit_of_measurement = unit_of_measurement
-        self._sensorid = sensorid
-
+        self._attr_unique_id = f"{entry_id}-{elem[FEED_ID]}"
```
we now use entry_id here. But we had something unique from the service right? But they needed a minimum required version. Should we use that unique_id here, and if the user did not upgrade yet to the latest, we use the entry_id. Once we see they updated, we can run a migration to change the unique_ids to the new unique_id

alexandrecuer replied:
No we do not have something unique from emoncms yet. I had initially included that kind of migration and you asked to keep that for a future PR. That's why I use the entry_id.

joostlek replied:
It's time for weekend lol, okay, let's do this in a followup

joostlek commented on homeassistant/components/emoncms/strings.json:
```diff
@@ -0,0 +1,21 @@
+{
+  "config": {
+    "step": {
+      "user": {
+        "data": {
+          "url": "[%key:common::config_flow::data::url%]",
+          "api_key": "[%key:common::config_flow::data::api_key%]"
+        }
+      }
```
Please add context using data_description

alexandrecuer replied:
done

joostlek reviewed (changes requested):
Small comments, will look at the tests later

alexandrecuer committed:
Define the type explicitely

alexandrecuer committed:
Add data description in strings.json

joostlek commented on tests/components/emoncms/conftest.py:
```diff
@@ -29,16 +34,43 @@ def get_feed(
 
 EMONCMS_FAILURE = {"success": False, "message": "failure"}
 
+FLOW_RESULT = {
+    CONF_API_KEY: "my_api_key",
+    CONF_ONLY_INCLUDE_FEEDID: [str(i + 1) for i in range(len(UNITS))],
+    CONF_URL: "http://1.1.1.1",
+}
```
Please move to `const.py`

joostlek commented on tests/components/emoncms/test_config_flow.py:
```diff
@@ -0,0 +1,117 @@
+}
+
+
+async def flow_import(hass: HomeAssistant, yaml: dict[str, Any]) -> ConfigFlowResult:
+    """Import of a yaml config."""
+    return await hass.config_entries.flow.async_init(
+        DOMAIN,
+        context={"source": SOURCE_IMPORT},
+        data=yaml,
+    )
```
I think it makes sense to inline this tbh

alexandrecuer replied:
removed and reintegrated in the tests

joostlek commented on tests/components/emoncms/test_config_flow.py:
```diff
@@ -0,0 +1,117 @@
+    )
+
+
+async def test_flow_import_all_feeds(
+    hass: HomeAssistant,
+    emoncms_client: AsyncMock,
+) -> None:
+    """YAML import of all feeds - success test."""
+    result = await flow_import(hass, YAML)
+    assert result["type"] is FlowResultType.CREATE_ENTRY
+    assert result["title"] == SENSOR_NAME
+    assert result["data"] == FLOW_RESULT
+
+
+YAML_INCL_FEED = copy.deepcopy(YAML)
+YAML_INCL_FEED[CONF_ONLY_INCLUDE_FEEDID] = [2]
+
+FLOW_RESULT_INCL_FEED = copy.deepcopy(FLOW_RESULT)
+FLOW_RESULT_INCL_FEED[CONF_ONLY_INCLUDE_FEEDID] = ["2"]
+
+
+async def test_flow_import_include_feeds(
+    hass: HomeAssistant,
+    emoncms_client: AsyncMock,
+) -> None:
+    """YAML import with included feed - success test."""
+    result = await flow_import(hass, YAML_INCL_FEED)
+    assert result["type"] is FlowResultType.CREATE_ENTRY
+    assert result["title"] == SENSOR_NAME
```
every test that ends with CREATE_ENTRY is required to mock the async_setupe_entry, otherwise it will try to startup the integration

alexandrecuer replied:
done

joostlek commented on tests/components/emoncms/test_config_flow.py:
```diff
@@ -0,0 +1,117 @@
+    """YAML import - failure test."""
+    emoncms_client.async_request.return_value = EMONCMS_FAILURE
+    result = await flow_import(hass, YAML)
+    assert result["type"] == FlowResultType.ABORT
```
reason?

alexandrecuer replied:
to test the `async_abort` in the `async_step_import`

joostlek replied:
Oh sorry, I was too short here.

Should we also test the reason why the flow was aborted?

alexandrecuer replied:
done

joostlek commented on tests/components/emoncms/test_config_flow.py:
```diff
@@ -0,0 +1,117 @@
+        result["flow_id"],
+        user_input=USER_INPUT,
+    )
+    await hass.async_block_till_done()
```
```suggestion
```

joostlek commented on tests/components/emoncms/test_config_flow.py:
```diff
@@ -0,0 +1,117 @@
+        user_input=USER_INPUT,
+    )
+    await hass.async_block_till_done()
+    assert result["type"] is FlowResultType.CREATE_ENTRY
+    assert result["data"] == CONFIG_ENTRY
```
When updating options flow, the options field of the result is changed, so we should check that

alexandrecuer replied:
should be done I think

joostlek commented on tests/components/emoncms/test_config_flow.py:
```diff
@@ -0,0 +1,117 @@
+    await setup_integration(hass, config_entry)
+    result = await hass.config_entries.options.async_init(config_entry.entry_id)
+    await hass.async_block_till_done()
+    assert result["errors"]["base"] == "failure"
+    assert result["type"] == FlowResultType.FORM
+    assert result["step_id"] == "init"
```
```suggestion
    assert result["errors"]["base"] == "failure"
    assert result["type"] is FlowResultType.FORM
    assert result["step_id"] == "init"
```
Let's repatch the function and test if it's able to recover

joostlek commented on tests/components/emoncms/test_config_flow.py:
I am missing a test that tests if it tried to import the same config. Also, I am missing the user flows

alexandrecuer replied:
test to import the same yaml added

joostlek replied:
The user flows are more important though :)

alexandrecuer replied:
done

joostlek commented on tests/components/emoncms/test_sensor.py:
```diff
@@ -59,12 +126,14 @@ async def test_coordinator_update(
-        assert state == snapshot(name=entity_id)
+    await setup_integration(hass, config_single_feed)
+
+    entity_entries = er.async_entries_for_config_entry(
+        entity_registry, config_single_feed.entry_id
+    )
+    for entity_entry in entity_entries:
+        state = hass.states.get(entity_entry.entity_id)
+        assert state == snapshot(name=entity_entry.entity_id)
```
use the snapshot_platform helper for this at it will also capture the entity itself

alexandrecuer replied:
done

joostlek commented:
Just as a heads up, with this progress I think we can make 2024.9. The beta cut happens at wednesday, so if we can make it before then, would be amazing

alexandrecuer committed:
Update tests/components/emoncms/test_config_flow.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

alexandrecuer committed:
Update tests/components/emoncms/test_config_flow.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

alexandrecuer committed:
Add test import same yaml conf + corrections

alexandrecuer committed:
Add test user flow

alexandrecuer committed:
Use data_description...

alexandrecuer committed:
Use snapshot_platform in test_sensor

alexandrecuer committed:
Merge branch 'dev' into config_flow

alexandrecuer committed:
Transfer all fixtures in conftest

alexandrecuer committed:
Add async_step_choose_feeds to ask flows to user

alexandrecuer committed:
Test abortion reason in test_flow_import_failure

alexandrecuer committed:
Add issue when value_template is i yaml conf

alexandrecuer committed:
make text more expressive in strings.json

alexandrecuer committed:
Add issue when no feed imported during migration.

joostlek commented on tests/components/emoncms/test_config_flow.py:
```diff
@@ -0,0 +1,143 @@
+        context={"source": SOURCE_IMPORT},
+        data=YAML,
+    )
+    assert result["type"] == FlowResultType.ABORT
```
```suggestion
    assert result["type"] is FlowResultType.ABORT
```

joostlek commented on tests/components/emoncms/test_config_flow.py:
```diff
@@ -0,0 +1,143 @@
+        context={"source": SOURCE_IMPORT},
+        data=YAML,
+    )
+    assert result["type"] == FlowResultType.ABORT
```
```suggestion
    assert result["type"] is FlowResultType.ABORT
```

joostlek committed:
Update tests/components/emoncms/test_config_flow.py

joostlek committed:
Update tests/components/emoncms/test_config_flow.py
